File: app/src/main/java/com/veganbeauty/admin/data/local/ProductConverters.java
package com.veganbeauty.admin.data.local;

import androidx.room.TypeConverter;
import com.veganbeauty.admin.data.local.entities.KeyIngredient;
import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class ProductConverters {
    @TypeConverter
    public String fromStringList(List<String> value) {
        if (value == null) {
            return "[]";
        }
        return new JSONArray(value).toString();
    }

    @TypeConverter
    public List<String> toStringList(String value) {
        if (value == null || value.isEmpty()) {
            return Collections.emptyList();
        }
        List<String> list = new ArrayList<>();
        try {
            JSONArray array = new JSONArray(value);
            for (int i = 0; i < array.length(); i++) {
                list.add(array.getString(i));
            }
        } catch (Exception e) {
            // Handle cases where the value is a simple comma-separated string instead of a JSON array
            String[] split = value.split(",");
            for (String item : split) {
                String trimmed = item.trim();
                if (!trimmed.isEmpty()) {
                    list.add(trimmed);
                }
            }
        }
        return list;
    }

    @TypeConverter
    public String fromKeyIngredients(List<KeyIngredient> value) {
        if (value == null) {
            return "[]";
        }
        JSONArray array = new JSONArray();
        try {
            for (KeyIngredient item : value) {
                JSONObject obj = new JSONObject();
                obj.put("name", item.getName());
                obj.put("description", item.getDescription());
                array.put(obj);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return array.toString();
    }

    @TypeConverter
    public List<KeyIngredient> toKeyIngredients(String value) {
        if (value == null || value.isEmpty()) {
            return Collections.emptyList();
        }
        List<KeyIngredient> list = new ArrayList<>();
        try {
            JSONArray array = new JSONArray(value);
            for (int i = 0; i < array.length(); i++) {
                JSONObject obj = array.getJSONObject(i);
                KeyIngredient ki = new KeyIngredient();
                ki.setName(obj.optString("name", ""));
                ki.setDescription(obj.optString("description", ""));
                list.add(ki);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return list;
    }
}

File: app/src/main/java/com/veganbeauty/admin/data/local/ProductConverters.kt
package com.veganbeauty.admin.data.local

import androidx.room.TypeConverter
import com.veganbeauty.admin.data.local.entities.KeyIngredient
import org.json.JSONArray
import org.json.JSONObject

class ProductConverters {
    @TypeConverter
    fun fromStringList(value: List<String>?): String {
        return value?.let { JSONArray(it).toString() } ?: "[]"
    }

    @TypeConverter
    fun toStringList(value: String?): List<String> {
        if (value.isNullOrEmpty()) return emptyList()
        val list = mutableListOf<String>()
        try {
            val array = JSONArray(value)
            for (i in 0 until array.length()) {
                list.add(array.getString(i))
            }
        } catch (e: Exception) {
            // Handle cases where the value is a simple comma-separated string instead of a JSON array
            return value.split(",").map { it.trim() }.filter { it.isNotEmpty() }
        }
        return list
    }

    @TypeConverter
    fun fromKeyIngredients(value: List<KeyIngredient>?): String {
        if (value == null) return "[]"
        val array = JSONArray()
        for (item in value) {
            val obj = JSONObject()
            obj.put("name", item.name)
            obj.put("description", item.description)
            array.put(obj)
        }
        return array.toString()
    }

    @TypeConverter
    fun toKeyIngredients(value: String?): List<KeyIngredient> {
        if (value.isNullOrEmpty()) return emptyList()
        val list = mutableListOf<KeyIngredient>()
        try {
            val array = JSONArray(value)
            for (i in 0 until array.length()) {
                val obj = array.getJSONObject(i)
                list.add(
                    KeyIngredient(
                        name = obj.optString("name", ""),
                        description = obj.optString("description", "")
                    )
                )
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return list
    }
}
